File: Java Desktop Runtime/jruntime/Navigator.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package jruntime;

import java.util.*;

/**
 *
 * @author Riju
 */
public class Navigator 
{
       private boolean isRunning = false;
        private ArrayList<Vector2> nav_points = new ArrayList<Vector2>();
        private int current_frame = 0;
        private GameObject_Scene baseObject;
        private int navigation_speed = 0;
        private float slope = 0f;
        private float deltaX = 0f, deltaY = 0f;
        private int pos_x = 0 , pos_y = 0;
        private int update_counter = 0, total_updates = 0;
        
        public Navigator(GameObject_Scene baseObject,int navigation_speed)
        {
            this.baseObject = baseObject;
            this.navigation_speed = navigation_speed;
        }

        public void addPoint(Vector2 point )
        {
            if (!isRunning) nav_points.add(point);
        }

        public void deletePoint(int count)
        {
            if (count > -1 && count < nav_points.size( ))
            {
                nav_points.remove(count);
            }
        }
        
        private void makePath(Vector2 begin_point , Vector2 end_point)
        {
            deltaY = (end_point.y - begin_point.y);
            deltaX = (end_point.x - begin_point.x);

            if (Math.abs(deltaX) >= Math.abs(deltaY)) slope = deltaY / deltaX;
            else slope = deltaX / deltaY;

            total_updates = (int)(((float)Math.sqrt((double)Math.pow(Math.abs(deltaX), 2) + Math.pow(Math.abs(deltaY), 2))) / navigation_speed);
            update_counter = 0;

            pos_x = begin_point.x;
            pos_y = begin_point.y;
        }
        
        public void start()
        {
            if (!isRunning)
            {
                isRunning = true;
                current_frame = 0;

                if (current_frame < nav_points.size( ) && current_frame + 1 < nav_points.size())
                {
                    makePath(nav_points.get(current_frame), nav_points.get(current_frame + 1)); 
                    baseObject.pos_x = pos_x;
                    baseObject.pos_y = pos_y;
                }
            }
        }

        public boolean isNavigating()
        {
            return isRunning;
        }

        public void stop()
        {
            isRunning = false;
        }
        
        public void cameraUpdatePoints(Vector2 pos)
        {
            for(int cntr = 0;cntr < nav_points.size( );cntr++)
            {
                nav_points.get(cntr).x -= pos.x;
                nav_points.get(cntr).y -= pos.y;
            }
        }
        
        public void update()
        {
             if (current_frame < nav_points.size( ) && current_frame + 1 < nav_points.size())
                if (this.baseObject.obj_instance.img != null)
                    if (update_counter == total_updates)
                    {
                        current_frame++;

                        if (current_frame < nav_points.size( ) && current_frame + 1 < nav_points.size( )) makePath(nav_points.get(current_frame), nav_points.get(current_frame + 1));
                        else stop();
                    }
                    else
                    {
                        if (Math.abs(deltaX) >= Math.abs(deltaY))
                        {
                            baseObject.pos_x = pos_x;
                            baseObject.pos_y = (int)(slope * (pos_x - nav_points.get(current_frame).x) + nav_points.get(current_frame).y);
                            pos_x += navigation_speed * ((deltaX < 0) ? -1 : (deltaX > 0) ? 1 : 0);
                        }
                        else
                        {
                            baseObject.pos_y = pos_y;
                            baseObject.pos_x = (int)(slope * (pos_y - nav_points.get(current_frame).y) + nav_points.get(current_frame).x);
                            pos_y += navigation_speed * ((deltaY < 0) ? -1 : (deltaY > 0) ? 1 : 0);
                        }

                        update_counter++;
                    }
                else ;
            else stop();
        }
}

File: Windows Native Runtime/Navigator.cpp
#include "Navigator.h"
#include<math.h>

Navigator::Navigator(GameObject_Scene * baseObject, int navigation_speed)
{
	this->baseObject = baseObject;
	this->navigation_speed = navigation_speed;
}

void Navigator::addPoint(Vector2 point)
{
	if (!isRunning) nav_points.push_back(point);
}

void Navigator::deletePoint(int count)
{
	if (count > -1 && count < nav_points.size( ))
	{
		int cntr = 0;

		for (std::list<Vector2>::iterator cur_point = nav_points.begin(); cur_point != nav_points.end(); cur_point++,cntr++)
		{
			if (cntr == count)
			{
				this->nav_points.erase(cur_point);
			}
		}
	}
}

void Navigator::start()
{
	if (!isRunning)
	{
		isRunning = true;
		current_frame = 0;

		if (current_frame < nav_points.size( ))
		{
			int cntr = 0;
			Vector2 point;

			for (std::list<Vector2>::iterator cur_point = nav_points.begin(); cur_point != nav_points.end(); cur_point++, cntr++)
			{
				if (cntr == current_frame)
				{
					point = *cur_point;
					break;
				}
			}

			if (point.x - baseObject->pos_x == 0)
			{
				isVertical = true;
			}
			else
			{
				deltaX = point.x - baseObject->pos_x;
				deltaY = point.y - baseObject->pos_y;
				delta_error = abs(point.y - baseObject->pos_y / point.x - baseObject->pos_x);
				isVertical = false;
				error = 0;
			}
		}
	}
}

bool Navigator::isNavigating()
{
	return isRunning;
}

void Navigator::stop()
{
	isRunning = false;
}

void Navigator::update()
{
	if (current_frame < nav_points.size( ))
	{
		if (this->baseObject->obj_instance.img.baseImage != NULL)
		{
			int cntr = 0;
			Vector2 point;

			for (std::list<Vector2>::iterator cur_point = nav_points.begin(); cur_point != nav_points.end(); cur_point++, cntr++)
			{
				if (cntr == current_frame)
				{
					point = *cur_point;
					break;
				}
			}

			if (baseObject->pos_x + (baseObject->obj_instance.img.Width / 2) > point.x && baseObject->pos_x < point.x + 10 && baseObject->pos_y + (baseObject->obj_instance.img.Height / 2) > point.y && baseObject->pos_y < point.y + 10)
			{
				current_frame++;

				if (current_frame < nav_points.size( ))
				{
					if (point.x - baseObject->pos_x == 0)
					{
						isVertical = true;
					}
					else
					{
						deltaX = point.x - baseObject->pos_x;
						deltaY = point.y - baseObject->pos_y;
						delta_error = abs(point.y - baseObject->pos_y / point.x - baseObject->pos_x);
						isVertical = false;
						error = 0;
					}
				}
			}
			else
			{
				if (!isVertical)
				{
					if (error >= 0.5f)
					{
						baseObject->pos_y += (point.y < baseObject->pos_y) ? -navigation_speed : navigation_speed;
						error /= (deltaX + navigation_speed * deltaY * deltaX + deltaY + navigation_speed);
					}
					else
					{
						baseObject->pos_x += (point.x < baseObject->pos_x) ? -navigation_speed : navigation_speed;
						error += delta_error;
					}
				}
				else
				{
					baseObject->pos_y += (point.y < baseObject->pos_y) ? -navigation_speed : navigation_speed;
				}
			}
		}
	}
	else
	{
		stop();
	}
}